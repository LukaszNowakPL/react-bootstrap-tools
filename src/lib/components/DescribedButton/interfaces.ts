export interface DescribedButtonProps {
    description?: string;
    buttonType?: string;
    buttonName?: string;
}

export interface ButtonProps {
    buttonType?:string,
    buttonName?:string,
}

export interface Aaa {
    description?: string,
}