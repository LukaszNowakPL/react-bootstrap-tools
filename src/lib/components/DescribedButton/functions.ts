import { ButtonProps, DescribedButtonProps } from "./interfaces";
import { DescribedComponentProps } from "../DescribedComponent/interfaces";
import { GlyphiconProps } from "../Glyphicon/interfaces";

/**
 * Returns props used by Button component
 * @param props
 */
export const getButtonProps = (props: DescribedButtonProps) => {
  const buttonProps: ButtonProps = {};

  if (props.name) buttonProps.name = props.name;
  if (props.type) buttonProps.type = props.type;
  if (!!props.active || props.active === false)
    buttonProps.active = props.active;
  if (!!props.block || props.block === false) buttonProps.block = props.block;
  if (!!props.disabled || props.disabled === false)
    buttonProps.disabled = props.disabled;
  if (props.href) buttonProps.href = props.href;
  if (props.href && props.hrefNewWindow) buttonProps.target = "_blank";
  if (props.size) buttonProps.size = props.size;
  if (props.variant) buttonProps.variant = props.variant;
  if (props.buttonClassName) buttonProps.className = props.buttonClassName;
  if (props.id) buttonProps.id = props.id;
  if (props.bsPrefix) buttonProps.bsPrefix = props.bsPrefix;

  if (props.onClick) {
    if (
      props.onClickArg ||
      props.onClickArg === false ||
      props.onClickArg === 0 ||
      props.onClickArg === null
    ) {
      buttonProps.onClick = ev => props.onClick(props.onClickArg, ev);
    } else {
      buttonProps.onClick = props.onClick;
    }
  }

  return buttonProps;
};

/**
 * Returns props used by DescribedComponent
 * @param props
 */
export const getDescribedComponentProps = (props: DescribedButtonProps) => {
  const describedComponentProps: DescribedComponentProps = {
    children: null
  };

  if (props.description)
    describedComponentProps.description = props.description;
  if (props.className) describedComponentProps.className = props.className;
  if (props.placement) describedComponentProps.placement = props.placement;
  if (props.delay) describedComponentProps.delay = props.delay;
  if (!!props.defaultShow || props.defaultShow === false)
    describedComponentProps.defaultShow = props.defaultShow;
  if (props.trigger) describedComponentProps.trigger = props.trigger;

  return describedComponentProps;
};

/**
 * Returns props used by Glyphicon component
 * @param props
 */
export const getGlyphiconProps = (props: DescribedButtonProps) => {
  const glyphiconProps: GlyphiconProps = {
    icon: props.icon || ""
  };

  if (props.iconClassName) glyphiconProps.className = props.iconClassName;

  return glyphiconProps;
};
