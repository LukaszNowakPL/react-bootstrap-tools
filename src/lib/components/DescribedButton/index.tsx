import * as React from "react";
import { ButtonProps, DescribedButtonProps } from "./interfaces";
import DescribedComponent from "../DescribedComponent/index";
import { DescribedComponentProps } from "../DescribedComponent/interfaces";
import {
  getButtonProps,
  getDescribedComponentProps,
  getGlyphiconProps
} from "./functions";
import { Button } from "react-bootstrap";
import Glyphicon from "../Glyphicon";
import { GlyphiconProps } from "../Glyphicon/interfaces";

const GlyphComponent = (props: DescribedButtonProps) => {
  const glyphiconProps: GlyphiconProps = getGlyphiconProps(props);
  return <Glyphicon {...glyphiconProps} />;
};

const ButtonComponent = (props: DescribedButtonProps) => {
  const buttonProps: ButtonProps = getButtonProps(props);

  const { icon, text } = props;

  const GlyphiconComponent = icon ? GlyphComponent(props) : null;

  return (
    <Button {...buttonProps}>
      {GlyphiconComponent}
      {text}
    </Button>
  );
};

const DescribedButton = (props: DescribedButtonProps) => {
  const { description } = props;

  const buttonComponent = ButtonComponent(props);

  if (description) {
    const describedComponentProps: DescribedComponentProps = getDescribedComponentProps(
      props
    );

    return (
      <DescribedComponent {...describedComponentProps}>
        {buttonComponent}
      </DescribedComponent>
    );
  }

  return buttonComponent;
};

export default DescribedButton;
