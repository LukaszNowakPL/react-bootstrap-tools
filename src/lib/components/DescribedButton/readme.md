# DescribedButton

Wrapper for describing Bootstrap's `Button` using `Tooltip` component. It also handles click actions, auto activation/disabling of button as well as simulates `href` link behaviour.

![DescribedButton example](../../../../readme_img/DescribedButtonExample.png)

## Example of usage

```
import React from 'react';
import { DescribedButton } from 'react-bootstrap-tools';

const SimplestExample = () => (
    <DescribedButton description="description text" text="button text" />
);

export default SimplestExample;
```

## Available optional props

- string `description` : Description text to put inside `Tooltip` component
- string `text` : Text visible on button
- string `icon` : Material icon id. If used together with `text` prop, icon will appear on the left hand side of text provided. Click [here](https://material.io/tools/icons/?style=baseline) to check available icons library.<br />
  Additional props connected with `icon` prop:
  - string `iconClassName` : Additional class name to custom your styling of provided icon
- string `className` : Additional class name to custom your styling of a `Tooltip` component.
  Styling example for `NewStyle` value:<pre>
  .NewStyle > .tooltip-inner {
  background-color: yellow;
  color: black;
  }
  .NewStyle.bs-tooltip-top .arrow::before {
  border-top-color: yellow !important;
  }
  .NewStyle.bs-tooltip-right .arrow::before {
  border-right-color: yellow !important;
  }
  .NewStyle.bs-tooltip-bottom .arrow::before {
  border-bottom-color: yellow !important;
  }
  .NewStyle.bs-tooltip-left .arrow::before {
  border-left-color: yellow !important;
  }</pre>
- string `buttonClassName` : Additional class name to custom your styling of button
- string `placement` : Position of a `Tooltip` over button.<br />
  Available options:
  - auto-start
  - auto
  - auto-end
  - top-start
  - top
  - top-end
  - right-start
  - right
  - right-end
  - bottom-end
  - bottom
  - bottom-start
  - left-end
  - left
  - left-start
- number `delay` : Number of milliseconds the `Tooltip` component will appear and hide after button hover.
- boolean `defaultShow` : Set `true` will initialise button with visible description
- string `trigger` : Action necessary to take over button to initialise visibility of description.<br />
  Available options:
  - hover
  - click
  - focus
  - Array<hover, click, focus>
- string `id` : Id property of rendered button
- string `name` : Name property of rendered button
- string `type` : Type property of rendered button.<br />
  Available options:
  - button
  - reset
  - submit
- boolean `active` : Set `true` will show button in visual `active` mode (same as it was activated by trigger action)
- boolean `block` : Set `true` will create block level buttons. It will span the full width of a parent
- boolean `disabled` : Set `true` will disable the button. This way the `onClick` action will not initialise
- string `href` : If on use, button will work as usual link directing to prop's url<br />
  Additional props connected with `href` prop:
  - boolean `hrefNewWindow` : Set `true` will open `href` url on new window<br />
- string `size` : Bootstrap's size property of a button.<br />
  Available options:
  - sm
  - lg
- string `variant` : Bootstrap's variant property of a button.<br />
  Available options:
  - primary
  - secondary
  - success
  - danger
  - warning
  - info
  - dark
  - light
  - link
  - outline-primary
  - outline-secondary
  - outline-success
  - outline-danger
  - outline-warning
  - outline-info
  - outline-dark
  - outline-light
- callback `onClick` : Callback function taken during click on button. The callback will receive `event` object that initiated the call.<pre>const onClickCallback = event => { /\*callback definition\*/ }</pre><br />
  Additional props connected with `onClick` prop:
  - any `onClickArg` : If on use, `callback` function will additionally receive this prop's data.<br /><pre>const onClickCallback = (customArg, event) => { /\*callback definition\*/ }</pre>This option is vital for better performance in terms of avoiding the unnecessary re-render of a component.