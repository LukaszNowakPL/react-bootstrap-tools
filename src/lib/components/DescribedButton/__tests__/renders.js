import React from "react";
import { shallow } from "enzyme";
import DescribedButton from "../index";

describe("<DescribedButton />", () => {
  const customComponentProps = {
    description: "test description"
  };

  it("renders", () => {
    const component = shallow(<DescribedButton {...customComponentProps} />);
    expect(component).toMatchSnapshot();
  });

  it("renders with custom button name", () => {
    const component = shallow(
      <DescribedButton {...customComponentProps} name={"testValue"} />
    );
    expect(component).toMatchSnapshot();
  });

  it("renders with custom button type", () => {
    const component = shallow(
      <DescribedButton {...customComponentProps} type={"submit"} />
    );
    expect(component).toMatchSnapshot();
  });

  it("renders with custom active state", () => {
    const component = shallow(
      <DescribedButton {...customComponentProps} active={true} />
    );
    expect(component).toMatchSnapshot();
  });

  it("renders with custom block", () => {
    const component = shallow(
      <DescribedButton {...customComponentProps} block={true} />
    );
    expect(component).toMatchSnapshot();
  });

  it("renders with custom disabled state", () => {
    const component = shallow(
      <DescribedButton {...customComponentProps} disabled={true} />
    );
    expect(component).toMatchSnapshot();
  });

  it("renders as link opening on new window", () => {
    const component = shallow(
      <DescribedButton
        {...customComponentProps}
        href={"testValue"}
        hrefNewWindow={true}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it("renders with custom size", () => {
    const component = shallow(
      <DescribedButton {...customComponentProps} size={"sm"} />
    );
    expect(component).toMatchSnapshot();
  });

  it("renders with custom variant", () => {
    const component = shallow(
      <DescribedButton {...customComponentProps} variant={"secondary"} />
    );
    expect(component).toMatchSnapshot();
  });

  it("renders with custom className of button", () => {
    const component = shallow(
      <DescribedButton
        {...customComponentProps}
        buttonClassName={"testValue"}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it("renders with custom text", () => {
    const component = shallow(
      <DescribedButton {...customComponentProps} text={"testValue"} />
    );
    expect(component).toMatchSnapshot();
  });

  it("renders with custom icon", () => {
    const component = shallow(
      <DescribedButton {...customComponentProps} icon={"testValue"} />
    );
    expect(component).toMatchSnapshot();
  });

  it("renders with custom icon and text", () => {
    const component = shallow(
      <DescribedButton
        {...customComponentProps}
        icon={"testIcon"}
        text={"testText"}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it("renders with custom className of an icon", () => {
    const component = shallow(
      <DescribedButton
        {...customComponentProps}
        icon={"testIcon"}
        iconClassName={"testValue"}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it("renders with custom onClick action with given argument", () => {
    const onClickMock = jest.fn();
    const onClickArg = { testKey: "test value" };
    const component = shallow(
      <DescribedButton
        {...customComponentProps}
        onClick={onClickMock}
        onClickArg={onClickArg}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it("renders with custom button id", () => {
    const component = shallow(
      <DescribedButton {...customComponentProps} id={"testValue"} />
    );
    expect(component).toMatchSnapshot();
  });

  it("renders with no description", () => {
    const component = shallow(<DescribedButton />);
    expect(component).toMatchSnapshot();
  });

  it("renders with custom Tooltip className", () => {
    const component = shallow(
      <DescribedButton {...customComponentProps} className={"testValue"} />
    );
    expect(component).toMatchSnapshot();
  });

  it("renders with custom Tooltip placement", () => {
    const component = shallow(
      <DescribedButton {...customComponentProps} placement={"top"} />
    );
    expect(component).toMatchSnapshot();
  });

  it("renders with custom Tooltip delay", () => {
    const component = shallow(
      <DescribedButton {...customComponentProps} delay={3000} />
    );
    expect(component).toMatchSnapshot();
  });

  it("renders with Tooltip show from beginning", () => {
    const component = shallow(
      <DescribedButton {...customComponentProps} defaultShow={true} />
    );
    expect(component).toMatchSnapshot();
  });

  it("renders with custom Tooltip trigger", () => {
    const component = shallow(
      <DescribedButton {...customComponentProps} trigger={"hover"} />
    );
    expect(component).toMatchSnapshot();
  });
});
