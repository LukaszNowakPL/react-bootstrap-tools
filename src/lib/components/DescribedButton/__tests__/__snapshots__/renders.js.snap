// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DescribedButton /> renders 1`] = `
<DescribedComponent
  description="test description"
>
  <ForwardRef(Bootstrap(Button)) />
</DescribedComponent>
`;

exports[`<DescribedButton /> renders as link opening on new window 1`] = `
<DescribedComponent
  description="test description"
>
  <ForwardRef(Bootstrap(Button))
    href="testValue"
    target="_blank"
  />
</DescribedComponent>
`;

exports[`<DescribedButton /> renders with Tooltip show from beginning 1`] = `
<DescribedComponent
  defaultShow={true}
  description="test description"
>
  <ForwardRef(Bootstrap(Button)) />
</DescribedComponent>
`;

exports[`<DescribedButton /> renders with custom Tooltip className 1`] = `
<DescribedComponent
  className="testValue"
  description="test description"
>
  <ForwardRef(Bootstrap(Button)) />
</DescribedComponent>
`;

exports[`<DescribedButton /> renders with custom Tooltip delay 1`] = `
<DescribedComponent
  delay={3000}
  description="test description"
>
  <ForwardRef(Bootstrap(Button)) />
</DescribedComponent>
`;

exports[`<DescribedButton /> renders with custom Tooltip placement 1`] = `
<DescribedComponent
  description="test description"
  placement="top"
>
  <ForwardRef(Bootstrap(Button)) />
</DescribedComponent>
`;

exports[`<DescribedButton /> renders with custom Tooltip trigger 1`] = `
<DescribedComponent
  description="test description"
  trigger="hover"
>
  <ForwardRef(Bootstrap(Button)) />
</DescribedComponent>
`;

exports[`<DescribedButton /> renders with custom active state 1`] = `
<DescribedComponent
  description="test description"
>
  <ForwardRef(Bootstrap(Button))
    active={true}
  />
</DescribedComponent>
`;

exports[`<DescribedButton /> renders with custom block 1`] = `
<DescribedComponent
  description="test description"
>
  <ForwardRef(Bootstrap(Button))
    block={true}
  />
</DescribedComponent>
`;

exports[`<DescribedButton /> renders with custom button id 1`] = `
<DescribedComponent
  description="test description"
>
  <ForwardRef(Bootstrap(Button))
    id="testValue"
  />
</DescribedComponent>
`;

exports[`<DescribedButton /> renders with custom button name 1`] = `
<DescribedComponent
  description="test description"
>
  <ForwardRef(Bootstrap(Button))
    name="testValue"
  />
</DescribedComponent>
`;

exports[`<DescribedButton /> renders with custom button type 1`] = `
<DescribedComponent
  description="test description"
>
  <ForwardRef(Bootstrap(Button))
    type="submit"
  />
</DescribedComponent>
`;

exports[`<DescribedButton /> renders with custom className of an icon 1`] = `
<DescribedComponent
  description="test description"
>
  <ForwardRef(Bootstrap(Button))>
    <Glyphicon
      className="testValue"
      icon="testIcon"
    />
  </ForwardRef(Bootstrap(Button))>
</DescribedComponent>
`;

exports[`<DescribedButton /> renders with custom className of button 1`] = `
<DescribedComponent
  description="test description"
>
  <ForwardRef(Bootstrap(Button))
    className="testValue"
  />
</DescribedComponent>
`;

exports[`<DescribedButton /> renders with custom disabled state 1`] = `
<DescribedComponent
  description="test description"
>
  <ForwardRef(Bootstrap(Button))
    disabled={true}
  />
</DescribedComponent>
`;

exports[`<DescribedButton /> renders with custom icon 1`] = `
<DescribedComponent
  description="test description"
>
  <ForwardRef(Bootstrap(Button))>
    <Glyphicon
      icon="testValue"
    />
  </ForwardRef(Bootstrap(Button))>
</DescribedComponent>
`;

exports[`<DescribedButton /> renders with custom icon and text 1`] = `
<DescribedComponent
  description="test description"
>
  <ForwardRef(Bootstrap(Button))>
    <Glyphicon
      icon="testIcon"
    />
    testText
  </ForwardRef(Bootstrap(Button))>
</DescribedComponent>
`;

exports[`<DescribedButton /> renders with custom onClick action with given argument 1`] = `
<DescribedComponent
  description="test description"
>
  <ForwardRef(Bootstrap(Button))
    onClick={[Function]}
  />
</DescribedComponent>
`;

exports[`<DescribedButton /> renders with custom size 1`] = `
<DescribedComponent
  description="test description"
>
  <ForwardRef(Bootstrap(Button))
    size="sm"
  />
</DescribedComponent>
`;

exports[`<DescribedButton /> renders with custom text 1`] = `
<DescribedComponent
  description="test description"
>
  <ForwardRef(Bootstrap(Button))>
    testValue
  </ForwardRef(Bootstrap(Button))>
</DescribedComponent>
`;

exports[`<DescribedButton /> renders with custom variant 1`] = `
<DescribedComponent
  description="test description"
>
  <ForwardRef(Bootstrap(Button))
    variant="secondary"
  />
</DescribedComponent>
`;

exports[`<DescribedButton /> renders with no description 1`] = `<ForwardRef(Bootstrap(Button)) />`;
